platform/pulp/platform/tasks/task_system.py

```diff
@@ -1,4 +1,548 @@
+        :type task_id:  :class: `uuid.UUID`
+
+        :param args:    Original arguments for the executed task.
+        :type args:     list
+
+        :param kwargs:  Original keyword arguments for the executed task.
+        :type kwargs:   dict
+
+        :param einfo:   celery's ExceptionInfo instance, containing serialized traceback.
+        :type einfo:    ???
+        """
+        if isinstance(exc, PulpCodedException):
+            _logger.info(_('Task failed : [%(task_id)s] : %(msg)s') %
+                         {'task_id': task_id, 'msg': str(exc)})
+            _logger.debug(traceback.format_exc())
+        else:
+            _logger.info(_('Task failed : [%s]') % task_id)
+            # celery will log the traceback
+        if kwargs.get('scheduled_call_id') is not None:
+            utils.increment_failure_count(kwargs['scheduled_call_id'])
+        if not self.request.called_directly:
+            now = datetime.now(dateutils.utc_tz())
+            finished_at = dateutils.format_iso8601_datetime(now)
+            task_status = TaskStatus.objects.get(id=task_id)
+            task_status.state = TaskStatus.ERRORED
+            task_status.finished_at = finished_at
+            task_status.error = {'traceback': einfo.traceback}
+            if not isinstance(exc, PulpException):
+                exc = PulpException(str(exc))
+                task_status.error = exc.to_dict()
+
+            task_status.save()
+            common_utils.delete_working_directory()
+
+
+def cancel(task_id):
+    """
+    Cancel the task that is represented by the given task_id. This method cancels only the task
+    with given task_id, not the spawned tasks. This also updates task's state to 'canceled'.
+
+    :param task_id: The ID of the task you wish to cancel
+    :type  task_id: basestring
+
+    :raises MissingResource: if a task with given task_id does not exist
+    :raises PulpCodedException: if given task is already in a complete state
+    """
+    try:
+        task_status = TaskStatus.objects.get(id=task_id)
+    except TaskStatus.DoesNotExist:
+        raise MissingResource(task_id)
+
+    if task_status.state in constants.CALL_COMPLETE_STATES:
+        # If the task is already done, just stop
+        msg = _('Task [%(task_id)s] already in a completed state: %(state)s')
+        _logger.info(msg % {'task_id': task_id, 'state': task_status.state})
+        return
+
+    if task_status.worker.name == 'agent':
+        tag_dict = dict(
+            [
+                tags.parse_resource_tag(t) for t in task_status['tags'] if tags.is_resource_tag(t)
+            ])
+        agent_manager = managers.consumer_agent_manager()
+        consumer_id = tag_dict.get(tags.RESOURCE_CONSUMER_TYPE)
+        agent_manager.cancel_request(consumer_id, task_id)
+    else:
+        celery_controller.revoke(task_id, terminate=True)
+
+    qs = TaskStatus.objects.get(id=task_id, state__neq=constants.CALL_COMPLETE_STATES)
```

Comment:
Good point. This code has been reworked so this won't be an issue on the next push.